### ISSUE TITLE
Замечания по LineCounter

### ISSUE THREAD

**Issue by sfactor** [bukshev/nsudotnet#1]
Общие замечания:
- программа не должна ничего считывать со стандартного ввода: тип подсчитываемых файлов должен задаваться параметром командной строки (например, `LineCounter.exe -type *.cs`)
- программа слишком много выводит на экран: достаточно количества подсчитанных файлов и суммарного количества значимых строк в них.

Более мелкие замечания:
- Проверка args на null бессмысленна — там всегда будет ненулевая ссылка (в худшем случае — ссылка на пустой массив)
- Не `line.Length < 1`, а `String.isNullOrEmpty()`
- Алгоритм разбора строки вопиюще неэффективный: там и создание лишних объектов через `.Trim()`, и три поиска подстроки подряд, и два вызова `.Split()`. Всю эту логику вполне можно было бы реализовать за один проход по стоке.

Кроме того, сам алгоритм подсчета строк получился довольно запутанным. Я бы предложил написать на него небольшой test suite: 5-10 маленьких тестовых файлов с разными конфигурациями комментариев, и сделать так, чтобы из VS Solution можно было запустить этот набор тестов и посмотреть, какие из них успешно проходят.

**Comment by bukshev** [comment]
С замечаниями ознакомлен. 
А вот по поводу тестов вопрос: подойдёт ли оформление test suite в виде отдельного тестового модуля? 
Я создал 7 различных файлов и для каждого из них в модуле прописал "персональный" метод. Для тестирования остаётся лишь в обозревателе тестов нажать запуск. 
После прокрутки тестов можно убедиться в корректности программы. Возможно, где-то перемудрил и пихнул костыли, но, вроде бы, работает. :)

**Comment by sfactor** [comment]
Да, оформление нормальное, только я бы тестовые файлы для подсчета не оформлял бы как отдельные .txt, а просто пихал бы их в код тестового модуля— так тесты будет легче и писать, и читать, и проверять.

Перенеси их, а потом я и сами тесты проверю

**Comment by bukshev** [comment]
Сделано.

**Comment by sfactor** [comment]
Ну если уж совсем красиво к делу подходить, то можно было переделать архитектуру так, чтобы создавать отдельный текстовый файл на каждый тест не требовалось (например, в `GetStringCount()` передавать `TextReader`, а не файл, а каждый тест оформить как наследника `TextReader`. Но это уже сильные украшательства:)

И многострочные строки в C# можно задавать целиком

```
string long_string = @"Hello
World
Hello
World";
```

Тестами ты меня прям порадовал, но алгоритм анализа строки надо переделать, а то он уж очень неэффективный. Переделай, и я добавлю два балла.

**Comment by bukshev** [comment]
Переписал функцию подсчёта, используя регулярные выражения. 
По поводу тестов согласен, много переиспользования кода. Как говорится, лень делает всякое дело трудным :)

**Comment by sfactor** [comment]
Ну блин:) Регэкспы ты добавил, это хорошо. Но всякая фигня типа `IndexOf()` и даже `string.Concat()` осталась! 

Неужели непонятно, что она добавляет сложности? 

Да, многострочные комментарии не парсятся регэкспом, то их можно было обработать **одним** проходом по строке.

Короче, ставлю два балла авансом, но завтра подойди и поговорим про это решение.
